cli.js

```diff
@@ -1,84 +1,132 @@
+    await handleCommand(argv, createHandler);
+  })
+  .command('deploy <function> [options]', 'Deploys a function to the cloud', (yargs0) => {
+    yargs0
+      .usage('Usage: $0 deploy <function> [options]')
+      .positional('function', {
+        describe: 'Name of the function to deploy',
+        type: 'string',
+      });
+  }, async (argv) => {
+    await handleCommand(argv, deployHandler);
+  })
+  .command('remove <function> [options]', 'Remove a previously deployed function', (yargs0) => {
+    yargs0
+      .usage('Usage: $0 remove <function> [options]')
+      .positional('function', {
+        describe: 'Name of the function to remove',
+        type: 'string',
+      });
+  }, async (argv) => {
+    await handleCommand(argv, removeHandler);
+  })
+  .command('invoke <function> [options]', 'Invoke a Binaris function', (yargs0) => {
+    yargs0
+      .usage('Usage: $0 invoke <function> [options]')
+      .positional('function', {
+        describe: 'Name of the function to invoke',
+        type: 'string',
+      })
+      .option('json', {
+        alias: 'j',
+        describe: 'Path to file containing JSON data',
+        type: 'string',
+      })
+      .option('data', {
+        alias: 'd',
+        describe: 'Data to send with invocation',
+        type: 'string',
+      });
+  }, async (argv) => {
+    await handleCommand(argv, invokeHandler);
+  })
+  .command('logs <function> [options]', 'Print the logs of a function', (yargs0) => {
+    yargs0
+      .usage('Usage: $0 logs <function> [options]')
+      .positional('function', {
+        describe: 'Name of the function to retrieve logs for',
+        type: 'string',
+      })
+      .option('tail', {
+        alias: 't',
+        describe: 'Outputs logs in "tail -f" fashion',
+        type: 'boolean',
+      });
+  }, async (argv) => {
+    await handleCommand(argv, logsHandler);
+  })
+  .command('login', 'Login to your Binaris account using an API key', (yargs0) => {
+    yargs0
+      .usage('Usage: $0 login');
+  }, async () => {
+    await loginHandler();
+  })
+  // .strict()
+  .demand(1, 'Please provide at least 1 valid command')
+  .help('help')
+  .alias('help', 'h')
+  .wrap(null);
 
-commander
-  .command('remove <function>')
-  .description('Remove a previously deployed function')
-  .action(namedActionHandler(removeHandler));
+const commands = yargs.getCommandInstance().getCommands();
+const argv = yargs.argv;
 
-commander
-  .command('invoke <function>')
-  .description('Invoke a Binaris function')
-  .option('-d, --data <data>', 'Data to send with invocation')
-  .option('-j, --json <filePath>', 'Path to file containing JSON data')
-  .action(namedActionHandler(invokeHandler));
-
-commander
-  .command('*', null, { noHelp: true })
-  .description('')
-  .action(actionWrapper(unknownHandler));
-
-commander
-  .parse(process.argv);
-
-if (!process.argv.slice(2).length) {
-  commander.outputHelp();
+if (argv._[0] && commands.indexOf(argv._[0]) === -1) {
```

Comment:
Highly unreadable to me. Looks like perl. wtf is it even doing?

Reply:
I agree it's semi-nasty but it's the canonical yargs way of handling this. This is a fall through check in case a command isn't handled by any of the command handlers. `argv._` holds all of the commands in the current yarg pipeline. That means that `argv._[0]` always holds the first command sent by the user. Then we check versus the master list of valid commands and if it isn't present we inform the user.